### ISSUE TITLE
MySQL duplicate and missing flac files - how best to fix with gui

### ISSUE THREAD

**Issue by brianb22** [gerbera/gerbera#514]
My media server runs ubuntu server bionic 18.04.3LTS with the 4.15.0 generic amd64 kernel. It uses the gerbera 1.3.2-1-ppa-bionic deb created by Stephen Czetty.

I have a large music collection, mostly with flac files which have simple metadata created by easyTag. Occasionally I decide to retag an album to suit my search preferences. I normally delete the old flac files, retain the directory, then move the new versions to the same place. This process updates the file timestamp.

The whole collection has an autoscan interval of 3 hours, so I usually just wait for the database to be updated. If I am in a hurry, I will reboot the system to be sure both gerbera and mysql start properly.

Lately this process has not been working as intended. I don't think anything has changed, so perhaps I am more observant these days. Often the problem is due to a track having the same value for ARTIST and ALBUM ARTIST. My update would completely remove the ALBUM ARTIST metadata. Another problem can arise when I change the ARTIST value, e.g. change "Who" to "The Who".

I primarily use the Yamaha DLNA implementations (Android MusiCast app with associated hardware). I also verify the track metadata with the gerbera web gui and independently with ffprobe.

I have read issue 401 carefully, and decided it was worth implementing the changes to config.xml to match the ubuntu mime.types (i.e. add `<treat mimetype="audio/flac" as="flac"/>` and `<map from="flac" to="audio/flac"/>`). This does not seem to have made any difference.

Previously, I have tried using the gui to delete every second track entry. Sometimes, my change "sticks", but I've also seen the duplicate return soon after (perhaps a result of an autoscan?). A rename of the ALBUM tags corrected the album entries in the gui display, while the artist entries lost most of the tracks. This mis-match relates to the same physical flac files!

I am reluctant to wipe the entire database and allow it to repopulate. This is because I know there are quite a few more albums and tracks which I want to retag in the future. I feel I ought to be able to re-tag an album and either have autoscan update the database properly, or at least be able to tidy it up once the improved-but-still-wrong change has propagated.

Here is an example of a re-tagged track which currently shows up as duplicated in both the album and artist indexes..

[gerbera-2019-10-25-mysql-bell-boy.TXT](https://github.com/gerbera/gerbera/files/3769962/gerbera-2019-10-25-mysql-bell-boy.TXT)

I'm not sure if this is helpful, but here is the related database content...

[gerbera-2019-10-25-mysql-bell-boy.TXT](https://github.com/gerbera/gerbera/files/3769963/gerbera-2019-10-25-mysql-bell-boy.TXT)

I would be grateful for any advice on how I can resolve my current problem, and also how to either avoid or resolve it as I make changes to my metadata in the future.

**Comment by brianb22** [comment]
I tried fixing some obvious problems with some variations of my procedures, but nothing worked properly. I stumbled over a clue - most of the duplicates I examined with EasyTag had the `ALBUM` and `ARTIST` tags, but also identical values for the `album_artist` tag. I quoted them case-sensitive, because ffprobe renders them both in capitals. I suspect the default gerbera index engine might be propagating the `album_artist` values into both the album and artist indexes.

Most of my efforts to correct the tags and replace the flacs in the music collection were dismal failures. In many cases only a fraction of the tracks in an album appeared on my Yamaha app, but ALL of them were duplicated!

I conclude this mysql database is completely corrupt. It was created with an older version of gerbera at the start of this year, so perhaps contains residual buggy data. I am going to throw it out and start with a clean copy.

**Comment by brianb22** [comment]
I have just finished creating an empty gerbera mysql database and defined autoscan for the top level of my media directory. After re-starting gerbera.service the database has been populated and many of the "wrong" entries are now correct.

With a good recipe for the work, it shouldn't be too difficult for me to do this all again. In the meantime, I need to understand why some of my metadata corrections seem to have been ignored. In particular, one entire album is still showing up under the unknown artist and album entries - using both the gerbera gui and my yamaha android app.

**Comment by whyman** [comment]
Thanks for the detailed bug report! Can you try it with a master (git) build? Or a 1.3.3 Build once @stephen-czetty kindly bumps the PPA?

I will let you know once Ive had a chance to look at the logs

**Comment by brianb22** [comment]
On 29/10/19 2:55 am, Ian Whyman wrote:
> Thanks for the detailed bug report!

Thanks for letting me know that someone cares! I've made quite a lot of 
progress in trying to simplify the issues so that I could "fix" the 
metadata which is causing the problems.

So far, I've done a lot of sifting, comparing and simplifying. However, 
the specific cases I am working on are baffling me because I can't 
detect any significant differences, and yet even with a 100% new 
database the albums are not processed properly.

Don't worry too much about working on my earlier examples at this stage, 
because I will be posting some much "simpler" (i.e. more baffling) cases 
very soon.

> Can you try it with a master (git) 
> build?

Before I worked with Stephen I had struggled and failed to build the 
project from source. There were simply too many pre-requisites and I 
couldn't fight my way through the build maze. Stephen has a build system 
for debian, so I was very happy to let him rescue me! Incidentally, his 
1.3.1 build fixed my problem at the time.

My own development work is mainly in java, so I am no stranger to adding 
many packages before I can build and debug problems. However, gerbera 
was beyond my experience and so fixing the build was very time-consuming 
and frustrating. I am reluctant to revive that task after my previous pain!

> Or a 1.3.3 Build once @stephen-czetty 
> <https://github.com/stephen-czetty> kindly bumps the PPA?

Oh PLEASE do, Stephen! I would be very grateful for another one of your 
rescue missions!

> I will let you know once Ive had a chance to look at the logs

As I said above, Ian, I don't want you spending a lot of time on the 
original problem report. I have much clearer examples I will post soon.

In the meantime, it is worth mentioning I have done some research into 
metadata. I had not previously realised there was such a great 
difference between flac and mp3 tagging. Gerbera and my other tools had 
hidden this important fact from me, but now I realise gerbera must have 
two different mapping engines to maintain its agnostic indexes.

My music collection is 98% flac and m4a (aac), so it isn't meaningful to 
rely on the fact that my few mp3's are processed properly, However, I've 
already stripped the metadata in my sample rogue flacs to the point 
where they /ought/ to map perfectly with the very limited set of ID3v1 
tags and values. Apparently that is not sufficient to resolve the problem!

I will update my post with these new examples soon, although it would be 
wise to delay debugging effort until I can test my collection against a 
more recent build.

Thanks very much again. I look forward with confidence to resolving my 
own indexing problems at the same time as improving such a valuable product.

Regards,

Brian

**Comment by brianb22** [comment]
Just as an appetiser, I will report on two almost identical flac files. I will attach their respective reports from metaflac for comparison. The other eight tracks in each album are similarly tagged, so it isn't surprising they are all indexed in the same manner as my two examples.

The first is properly indexed, e.g to Artists->Sade->Diamond Life and Albums->Diamond Life

[gerbera-2019-10-29-DiamondLife.txt](https://github.com/gerbera/gerbera/files/3781447/gerbera-2019-10-29-DiamondLife.txt)

The second is indexed in Artists->Unknown->Unknown and Albums->Unknown

[gerbera-2019-10-29-isItaCrime.txt](https://github.com/gerbera/gerbera/files/3781451/gerbera-2019-10-29-isItaCrime.txt)

The first looks a little suspect, but is indexed perfectly. The second looks clean and is mishandled! I have another pair of albums by James Taylor which exhibit the same kind of behaviour. I am happy to upload samples but there doesn't seem to be a lot of point until I have tested them against a newer version of gerbera.

**Comment by brianb22** [comment]
I cloned the master and went through the additional packages required (or optional) to run the build. I couldn't identify the correct ubuntu packages for taglib, libexif2, lastfmlib and inotify. I found candidates with similar names, but wasn't sure what to install. I found https://github.com/gerbera/gerbera/issues/100, but the discussion didn't help me enough.

I ran cmake without the missing optional packages, but it failed:-
`/home/brian/sandboxGerbera/build/CMakeFiles/CMakeTmp/src.cxx:2:13: fatal error: uuid.h: No such file or directory
    2 |    #include <uuid.h>
      |             ^~~~~~~~
compilation terminated.
`
It seems the latest version of uuid puts that file in <uuid/uuid.h>. I'm not sure how to change the build in a robust fashion. Of course, I could hack my own copy of FindUUID.cmake to get the correct path.

**Comment by brianb22** [comment]
No luck! That simple change just triggered several compilation errors, so I needed something more sophisticated.

[Oh yes, I forgot to mention I had to install uuid-dev to get the include file. I also had to install duktape-dev and libffmpegthumbnailer-dev for the same reason.]

cmake now runs successfully, even without my hack to uuid! I am a bit puzzled, but I'll take my luck where I find it...

The make and make install also ran OK. I now need to move the installed files to my media server....

**Comment by brianb22** [comment]
OOPS! I've just read the next section of the 1.4.0-beta manual, where it states very clearly which addition *-dev ubuntu packages are required for a successful build. Sorry for the noise!

**Comment by brianb22** [comment]
Phew! I transferred the 1.4 build from my 19.10 development system to my 18.04 production gerbera server. That didn't work because there were too many dependencies on different versions of shared libraries. I gave up and plodded through building it from scratch on the 18.04 system.

I got it to rebuild from a completely empty mysql media database. The original problem albums/tracks are still missing! (By the way, when I drill down from my yamaha android app via the "PC Directory" route, the tracks are all listed and play OK - but then they always did!)

I will copy a couple of sample tracks to my development system and confirm the problem is reproducible with the eoan shared libraries. At that time, I will need help to collect detailed diagnostic data.

**Comment by whyman** [comment]
Thanks for putting in the effort @brianb22!

**Comment by brianb22** [comment]
Good news in a strange way. I set up my ubuntu desktop 19.10 gerbera build from the 1.4-beta trunk with mysql server.

I added the four albums, two each from Sade and James Taylor. After they were indexed, the Albums and Artists entries shown on the gerbera gui exhibit exactly the same errors as on Stephen's 1.3.2 PPA.

I will be away from my office for most of this week, but I would be grateful for some advice on how to collect more diagnostic data when I return.

**Comment by whyman** [comment]
Can you provide a debug log?

**Comment by brianb22** [comment]
On 26/11/19 7:36 am, Ian Whyman wrote:
> Can you provide a debug log?

Yes, Ian. I have RTFM'd the beta version of the 1.4 manual (the colours 
don't render very clearly in the <code> blocks). I think you mean I 
should add the --debug switch to the command line.

I will recreate the music collection with only the four sample files 
(two good and two bad), then create a new mysql database. I hope that 
will provide you with more helpful diagnostics.

I should be back home soon and will put the work near the top of my 
priority list.

Regards,

Brian

**Comment by brianb22** [comment]
I've pinned down the problem, although I don't know how to circumvent it. Here are debug log extracts from two tracks:
1. Good metadata
`Dec  4 15:00:02 schizo gerbera[5816]: 2019-12-04 15:00:02   DEBUG: [src/content_manager.cc:771] addRecursive(): IS TASK VALID? [1], taskoath: [/home/media/Music/james-taylor/sweet]
Dec  4 15:00:02 schizo gerbera[5816]: 2019-12-04 15:00:02   DEBUG: [src/metadata/taglib_handler.cc:583] extractFLAC(): TagLibHandler: flac resource has no picture information
Dec  4 15:00:02 schizo gerbera[5816]: 2019-12-04 15:00:02   DEBUG: [src/metadata/taglib_handler.cc:241] fillMetadata(): TagLib handler done.
Dec  4 15:00:02 schizo gerbera[5816]: 2019-12-04 15:00:02   DEBUG: [src/metadata/ffmpeg_handler.cc:246] fillMetadata(): Running ffmpeg handler on /home/media/Music/james-taylor/sweet/01.Sweet_Baby_James.flac`
2. Bad Metadata
`Dec  4 15:00:06 schizo gerbera[5816]: 2019-12-04 15:00:06   DEBUG: [src/content_manager.cc:753] addRecursive(): Checking path /home/media/Music/james-taylor/gorilla
Dec  4 15:00:06 schizo gerbera[5816]: 2019-12-04 15:00:06   DEBUG: [src/content_manager.cc:771] addRecursive(): IS TASK VALID? [1], taskoath: [/home/media/Music/james-taylor/gorilla]
Dec  4 15:00:06 schizo gerbera[5816]: 2019-12-04 15:00:06   DEBUG: [src/metadata/taglib_handler.cc:547] extractFLAC(): TagLibHandler: could not open flac file: /home/media/Music/james-taylor/gorilla/01.Mexico.flac
Dec  4 15:00:06 schizo gerbera[5816]: 2019-12-04 15:00:06   DEBUG: [src/metadata/taglib_handler.cc:241] fillMetadata(): TagLib handler done.
Dec  4 15:00:06 schizo gerbera[5816]: 2019-12-04 15:00:06   DEBUG: [src/metadata/ffmpeg_handler.cc:246] fillMetadata(): Running ffmpeg handler on /home/media/Music/james-taylor/gorilla/01.Mexico.flac`

...and here are metaflac reports of the two tracks:
1. Good
`brian@schizo:/home/media/Music/james-taylor$ metaflac --list sweet/01.Sweet_Baby_James.flac 
METADATA block #0
  type: 0 (STREAMINFO)
  is last: false
  length: 34
  minimum blocksize: 4608 samples
  maximum blocksize: 4608 samples
  minimum framesize: 3989 bytes
  maximum framesize: 12907 bytes
  sample_rate: 44100 Hz
  channels: 2
  bits-per-sample: 16
  total samples: 7624596
  MD5 signature: 0b7f3159c352ff3293f0bca9107e5a98
METADATA block #1
  type: 4 (VORBIS_COMMENT)
  is last: false
  length: 163
  vendor string: Lavf57.71.100
  comments: 7
    comment[0]: TITLE=Sweet Baby James
    comment[1]: ARTIST=James Taylor
    comment[2]: ALBUM=Sweet Baby James
    comment[3]: DATE=1970
    comment[4]: TRACKNUMBER=01
    comment[5]: TRACKTOTAL=11
    comment[6]: GENRE=Folk/Rock
METADATA block #2
  type: 1 (PADDING)
  is last: true
  length: 8075`
2. Bad
`brian@schizo:/home/media/Music/james-taylor$ metaflac --list gorilla/01.Mexico.flac 
METADATA block #0
  type: 0 (STREAMINFO)
  is last: false
  length: 34
  minimum blocksize: 4608 samples
  maximum blocksize: 4608 samples
  minimum framesize: 5231 bytes
  maximum framesize: 13777 bytes
  sample_rate: 44100 Hz
  channels: 2
  bits-per-sample: 16
  total samples: 7699860
  MD5 signature: 817cb57a118c9267dfd9e4484279d5bf
METADATA block #1
  type: 3 (SEEKTABLE)
  is last: false
  length: 0
  seek points: 0
METADATA block #2
  type: 4 (VORBIS_COMMENT)
  is last: false
  length: 163
  vendor string: reference libFLAC 1.3.2 20170101
  comments: 7
    comment[0]: TITLE=Mexico
    comment[1]: ARTIST=James Taylor
    comment[2]: ALBUM=Gorilla
    comment[3]: DATE=1975
    comment[4]: TRACKNUMBER=01
    comment[5]: TRACKTOTAL=11
    comment[6]: GENRE=Folk/Rock
METADATA block #3
  type: 1 (PADDING)
  is last: true
  length: 321`

The good track has two vorbis metadata blocks, while the bad one has three. The debug indicates taglib_handler.cc:547 cannot process the flac file with the three metadata blocks and so all the important variables are assigned "Unknown" values.

Can I send you these two complete tracks, or have I already provided sufficient diagnostics for you to fix the bug?

**Comment by whyman** [comment]
Well I must say, that's excellent diagnostics indeed!

Unfortunately this doesnt appear to be an issue with Gerbera so much as TagLib: it seems like this issue has been raised with taglib a while ago and it seems that only a single comment is supported as that is what is defined in the FLAC spec.

https://github.com/taglib/taglib/issues/576#issuecomment-106084513
> According to the FLAC format: "There may be only one VORBIS_COMMENT block in a stream".

Gerbera simply asks TagLib if the file is valid via `flac.isValid()` and in your bad file it retuns false :man_shrugging:

**Comment by brianb22** [comment]
Initially, your response disappointed me. Then I remembered I only have this problem with two of my albums, both of which I digitised from vinyl a long time ago. I dug around and realised I used sox for the flac conversion in the early days, but moved to ffmpeg fairly quickly.

First I used EasyTag to remove all the tags from Mexico, but metatag showed me the first metadata block remained in the file. When I re-added the metadata, the file reverted to its "broken" layout.

Luckily, flac is lossless and so I used ffmpeg to convert the track back to a wav file, which obviously does not have any metadata. ffmpeg then compressed it back into a new flac without the metadata block that had caused all the trouble. After I added the correct metadata with EasyTag, gerbera created the correct indexes in its mysql database.

Problem solved! Thanks very much Ian.

**Comment by whyman** [comment]
Great news!

**Comment by brianb22** [comment]
I've been having a bad day... my procedure to create valid vorbis metadata didn't work on the two albums on my production media server. It seems my ffmpeg conversion from flac to wav assigns metadata to the wav file! I didn't know that was possible. The reverse conversion wav to flac simply recreates the bad metadata.

I'll have to try again tomorrow when my head is less cluttered.

**Comment by whyman** [comment]
Maybe use `flac -d` for a more stupid version?

It looks like `metaflac` (also in the `flac` package) has some useful sounding options namely:

```
--remove-all
    Remove all metadata blocks (except the STREAMINFO block) from the
    metadata.  Unless --dont-use-padding is specified, the blocks will be
    replaced with padding.

--merge-padding
    Merge adjacent PADDING blocks into single blocks.

--sort-padding
    Move all PADDING blocks to the end of the metadata and merge them into a
    single block.```

**Comment by brianb22** [comment]
In my two albums it was the seek table which was considered to be invalid. I found this command worked perfectly, so thanks for the tip!
`metaflac --remove --block-number=1 <trackname>`

I was wrong in thinking the error was due to the surplus seek table entries, because I subsequently discovered many more tracks have four metadata blocks, where the second (number 1) is seek data. I suppose those two broken albums must have been converted from wav files by sox which wrote invalid seek data blocks. I don't think we will ever know for sure.

However, my problem is definitely fixed this time! Thanks very much Ian.
